File: branches/tfs-dev-1.3.1-bugfix/src/message/rename_file_message.cpp
/*
 * (C) 2007-2010 Alibaba Group Holding Limited.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 *
 * Version: $Id$
 *
 * Authors:
 *   duolong <duolong@taobao.com>
 *      - initial release
 *
 */
#include "rename_file_message.h"

using namespace tfs::common;

namespace tfs
{
  namespace message
  {
    RenameFileMessage::RenameFileMessage() :
      option_flag_(0)
    {
      _packetHeader._pcode = RENAME_FILE_MESSAGE;
      memset(&rename_file_info_, 0, sizeof(RenameFileInfo));
    }

    RenameFileMessage::~RenameFileMessage()
    {
    }

    int RenameFileMessage::parse(char* data, int32_t len)
    {
      if (get_object_copy(&data, &len, reinterpret_cast<void*>(&rename_file_info_), sizeof(RenameFileInfo)) == TFS_ERROR)
      {
        return TFS_ERROR;
      }
      if (get_vint64(&data, &len, ds_) == TFS_ERROR)
      {
        return TFS_ERROR;
      }
      get_int32(&data, &len, &option_flag_);
      has_lease_ = parse_special_ds(ds_, version_, lease_id_);

      return TFS_SUCCESS;
    }

    int32_t RenameFileMessage::message_length()
    {
      int32_t len = sizeof(RenameFileInfo) + get_vint64_len(ds_) + INT_SIZE;
      if (has_lease_ == true)
      {
        len += INT64_SIZE * 3;
      }
      return len;
    }

    int RenameFileMessage::build(char* data, int32_t len)
    {
      if (has_lease_ == true)
      {
        ds_.push_back(ULONG_LONG_MAX);
        ds_.push_back(static_cast<uint64_t> (version_));
        ds_.push_back(static_cast<uint64_t> (lease_id_));
      }

      if (set_object(&data, &len, &rename_file_info_, sizeof(RenameFileInfo)) == TFS_ERROR)
      {
        return TFS_ERROR;
      }
      if (set_vint64(&data, &len, ds_) == TFS_ERROR)
      {
        return TFS_ERROR;
      }
      if (set_int32(&data, &len, option_flag_) == TFS_ERROR)
      {
        return TFS_ERROR;
      }

      has_lease_ = parse_special_ds(ds_, version_, lease_id_);

      return TFS_SUCCESS;
    }

    char* RenameFileMessage::get_name()
    {
      return "renamefilemessage";
    }

    Message* RenameFileMessage::create(const int32_t type)
    {
      RenameFileMessage* req_rf_msg = new RenameFileMessage();
      req_rf_msg->set_message_type(type);
      return req_rf_msg;
    }
  }
}

File: branches/tfs-test/tfsTest/src/main/java/com/taobao/common/tfs/packet/RenameFileMessage.java
/*
 * (C) 2007-2010 Alibaba Group Holding Limited.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 */
package com.taobao.common.tfs.packet;

import com.taobao.common.tfs.common.TfsConstant;

public class RenameFileMessage extends BasePacket {

    private int blockId;
    private long fileId;
    private long newFileId;
    private int isServer;
    private int optionFlag;
    private DsListWrapper dsListWrapper;

    public RenameFileMessage(Transcoder transcoder) {
        super(transcoder);
        this.pcode = TfsConstant.RENAME_FILE_MESSAGE;
    }

    @Override
    public int getPacketLength() {
        return TfsConstant.INT_SIZE * 3 + TfsConstant.LONG_SIZE * 2 + dsListWrapper.streamLength();
    }

    @Override
    public void writePacketStream() {
        byteBuffer.putInt(blockId);
        byteBuffer.putLong(fileId);
        byteBuffer.putLong(newFileId);
        byteBuffer.putInt(isServer);
        dsListWrapper.writeToStream(byteBuffer);
        byteBuffer.putInt(optionFlag);
    }

    public int getBlockId() {
        return blockId;
    }

    public void setBlockId(int blockId) {
        this.blockId = blockId;
    }

    public long getFileId() {
        return fileId;
    }

    public void setFileId(long fileId) {
        this.fileId = fileId;
    }

    public long getNewFileId() {
        return newFileId;
    }

    public void setNewFileId(long newFileId) {
        this.newFileId = newFileId;
    }

    public int getIsServer() {
        return isServer;
    }

    public void setIsServer(int isServer) {
        this.isServer = isServer;
    }

    public int getOptionFlag() {
        return optionFlag;
    }

    public void setOptionFlag(int optionFlag) {
        this.optionFlag = optionFlag;
    }

    public DsListWrapper getDsListWrapper() {
        return dsListWrapper;
    }

    public void setDsListWrapper(DsListWrapper dsListWrapper) {
        this.dsListWrapper = dsListWrapper;
    }


}
